[NJT] Store vec on nested ints

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117905
* #119977
* __->__ #119976
* #121481

This PR:
- adds data structure managing vec and nested int as well as metadata 
- recursively fakify tensors on symint

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @aakhundov

soulitzer committed:
[NJT] Store vec on nested ints

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Store vec on nested ints"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Store vec on nested ints"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Store vec on nested ints"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Store vec on nested ints"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented on torch/_subclasses/meta_utils.py:
```diff
@@ -241,6 +241,18 @@ def sym_sizes_strides_storage_offset(
+                def metafy_fn(t, src) -> torch.Tensor:
+                    _symbolic_context = None
+                    if symbolic_context is not None:
+                        _symbolic_context = symbolic_context.tensor_to_inner_context[t]
```
This doesn't actually work in this version of the data structure where we maintain a canonical nested int because the vec on the nested int won't necessarily the same vec as the offsets/lengths.

soulitzer replied:
One way to do this while keeping the current "canonical nested int" design is to maintain weak ref to all vec in the equiv set.

The alternative might be just to enforce that each vec must associated with its own nested int and enforce that NT's offset/lengths is always the same tensor as the vec in its nested int.

soulitzer replied:
I've updated the PR to do the "maintain a weak ref to every vec" approach. Still need to add a test.

soulitzer committed:
Update on "[NJT] Store vec on nested ints"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -3,23 +3,224 @@
+
+    def maybe_create(self, vec, *, ctor_fn=None, equiv_set_from=None, has_coeff=False):
+        # Given vec, return an associated nested int.
+        #
```
Add some discussion about what the user can expect from this API, e.g. the same vec implies nested int with same equiv id, but not necessarily the same instance. If the vec is later mutated, using it with the registry will error.

soulitzer committed:
Update on "[NJT] Store vec on nested ints"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -3,23 +3,224 @@
+            "Detected that vec has been mutated. This is not allowed. "
+        )
+
+    def maybe_create(self, vec, *, ctor_fn=None, equiv_set_from=None, has_coeff=False):
```
nit: I'd expect this to be named something about looking up the vec -> nested int in the registry, although I do appreciate that it indicates it may create a nested int

soulitzer replied:
Yeah happy to bikeshed on this a bit more - other options are probably `lookup`, `get`

jbschlosser replied:
maybe `lookup_or_create()` or something

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -3,23 +3,224 @@
+    #
+    def __init__(self):
+        self._equiv_set_counter = 0
+        self._equiv_sets = WeakTensorKeyDictionary()
```
nit: document the mapping modeled here. My understanding is it's something like `tensor weakref -> dict of properties associated with the equiv set` (including id, canonical vec, canonical nested int, weakrefs to all vecs, etc.)

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -3,23 +3,224 @@
+    # equiv set even though they are on different devices because they have the
+    # same data.
+    #
+    # The first vec and nested int associated with a equiv set becomes the
```
is it possible for there to be multiple nested ints associated with an equiv set? Naively, I thought there would be a 1:1 mapping between equiv set and nested int, with the possibility for multiple vecs within a single equiv set

soulitzer replied:
Yes, this happens in (at least?) the following situations:
- someone needs a nested int with a non-one coeff
- the existing nested int has been saved for backward, and the object in python dies; the next someone asks for a nested corresponding to the vec (which is still alive because the NestedIntSymNode still holds an owning ref)

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -3,23 +3,224 @@
+    # Pass a vec to maybe_create to create a new nested int. Multiple nested int
+    # can be associated with a vec. vec are then grouped into equivalence sets
+    # so that comparisons between the vec's nested ints can be made by comparing
+    # their equiv set. Morally two vec are in the same equiv set if and only if
+    # they have the same data, e.g., vec.cpu() and vec.cuda() belong to the same
+    # equiv set even though they are on different devices because they have the
+    # same data.
```
Is this moral behavior (same data -> same equiv set) enforced or is it the case that vecs are only part of the same equiv set if this is done explicitly? I think it's good to be clear on what's enforced or not here

soulitzer replied:
Not enforced, will update doc to make that clearer.

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -3951,6 +3954,142 @@ def in_proj(input_packed, qkv_linear=qkv_linear):
+            self.assertTrue(len(registry._equiv_sets) == 0)
+
+    def test_custom_equiv_set(self):
+        # A equiv set is "custom" if its canonical int was created using a
+        # custom ctor_fn.
```
what's the significance and purpose of a "custom" equiv set in practice?

soulitzer replied:
The main use is so that during compile, if I need a symbolic nested int, the creation of that can reuse the same logic as eager, for example enforcing that there is a single canonical vec/nested int

jbschlosser commented on aten/src/ATen/core/NestedIntSymNodeImpl.h:
```diff
@@ -179,8 +224,27 @@ class TORCH_API NestedIntSymNodeImpl : public SymNodeImpl {
+  // It should not be mutated
+  at::Tensor vec_;
+  // See NOTE [ NestedTensorVariant ]
+  NestedTensorVariant type_;
+  // Depending on the NestedTensorVariant we may want to either dispatch to
+  // Python or NestedTensor kernels. We dispatch to the Python kernel in the
+  // case the nested int corresponds to a NT python because we'd like for
+  // infra modes to be able to trace underneath.
+  c10::DispatchKeySet key_set_;
```
dumb Q: do we need both a key set and NestedTensorVariant enum? doesn't one imply the other?

soulitzer replied:
Good point, we can probably get rid of explicitly storing the key set

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -3,23 +3,224 @@
+        #        the equiv set corresponding to the vec specified by this arg.
+        #        The user is responsible for ensuring that vec and equiv_set_from
+        #        are compatible, i.e., they have the same data.
+        #     has_coeff (bool): If True, a new nested int will be created even
+        #        if vec is already in an equiv set. Note that we don't actually
+        #        allow coeff to passed to this function. The ctor_fn is expected
+        #        to handle this. Also note that this is not how nested ints with
+        #        coeff are typically created. Ordinarily, nested ints with coeff
+        #        are created by operations on existing nested ints.
```
nit: arg name and behavior don't match to me intuitively

soulitzer replied:
Yeah, the behavior is quite general, and the arg name hints at the specific and only use case it is used for. Let me try to see if I can make this make more sense.

jbschlosser reviewed (commented):
Thanks for all the hard work! There's a lot going on here so nice job in working through it :)

I'd benefit from a clear definition of "equiv set" (apologies if I missed it in the docs). Intuitively, I'd expect this to be a set of vecs that are all considered equivalent to each other and are associated with a single nested int. However, I think in practice the term is used slightly differently (certainly involving additional metadata), and I'd like to understand exactly how / all the metadata we expect to store related to the set.

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -3951,6 +3954,142 @@ def in_proj(input_packed, qkv_linear=qkv_linear):
+        # Vec is the only vec, and so it is also the canonical vec
+        self.assertIs(nested_int.node.nested_int_vec(), vec)
+        nested_int2 = registry.maybe_create(vec)
+        # Return the same nested int if same vec is passed in
```
right, should've been more clear, it needs to be the same object in python
technically any vec in the equivalent should result in the same nested int being returned, though today the user would've needed to manually specify that with the maybe_create(equiv_set_from=) API.

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -3951,6 +3954,142 @@ def in_proj(input_packed, qkv_linear=qkv_linear):
+        vec = torch.tensor([1, 2, 3])
+        vec_clone = vec.clone()
+        nested_int = registry.maybe_create(vec)
+        nested_int2 = registry.maybe_create(vec_clone, equiv_set_from=nested_int.node.nested_int_vec())
```
> just checking my understanding - it's required to explicitly indicate that a given vector is in an equivalence set with another vector?

that's exactly right, will make sure we have that test

soulitzer commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1572,20 +1572,23 @@ def _automatic_dynamic(
         # Get symbolic contexts for inner tensors
         attrs, _ = type(e).__tensor_flatten__(e)
         inner_contexts = {}  # mapping from attr -> symbolic context
+        tensor_to_inner_context = WeakTensorKeyDictionary()
```
> I'm guessing for recursive fake-ification of tensors associated with nested ints.

Yup, the issue here is that the fake tensor cache sees the offsets tensor on the nested int BEFORE it sees the offsets tensor during transform subclass. When we are fakifying the tensor on the nested int, we have no idea what attr it corresponds to, so we need some kind of mapping from tensor to context somehow.

> For simplicity, can we get away with using the pre-existing inner_contexts and building a mapping between attr <-> inner tensor on the fly during fake-ification? (or some other solution where we only need to store one thing related to inner contexts here)

Hmm right, storing a mapping from inner tensor to attr, and then reusing the attr to contexts mapping that we already have could work if that is what you mean.

soulitzer commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1299,10 +1300,25 @@ void initJITBindings(PyObject* module) {
+            TORCH_CHECK(node->is_nested_int());
             return node->nested_int_coeff();
+          })
+      .def(
+          "clone_nested_int_with_new_vec",
+          [](const c10::SymNode& node, const at::Tensor& vec) {
+            TORCH_CHECK(
+                false,
+                "Expect not to call clone_nested_int_with_new_vec on a "
+                "non-symbolic nested int");
```
This is just a stub to provide a nice error message if someone does nested_int.node.clone_nested_int_with_new_vec when the nested int is not symbolic. 

> and also is it overridden somehow for symbolic nested ints?

Yeah, the real logic here is actually in SymNode. 

The only reason this clone needs to exists is so that source is properly propagated and fx graph properly captures everything, so no one should be calling the clone here.

soulitzer commented on torch/csrc/utils/python_symnode.h:
```diff
@@ -100,6 +101,12 @@ class PythonSymNodeImpl : public c10::SymNodeImpl {
     return getPyObj().attr("is_nested_int")().is(py::handle(Py_True));
   }
 
+  c10::TensorImpl* nested_int_vec() const override {
+    TORCH_INTERNAL_ASSERT(
+        false,
+        "Expect not to call nested_int_vec on PythonSymNodeImpl in C++");
+  }
```
`class SymNode` has a method `nested_int_vec`. Usually methods on python_symnode.h would redirect there, but since the factory function creation logic is the only thing today that calls nested_int_vec and that is done in python from torch dispatch, there is no need to support this in cpp today.

soulitzer commented on torch/fx/experimental/sym_node.py:
```diff
@@ -181,6 +192,39 @@ def is_nested_int(self):
+        assert self._nested_int_vec is not None
+        return self._nested_int_vec
+
+    def clone_nested_int_with_new_vec(self, equiv_set, vec):
```
Not necessarily the same equiv set. There are actually two situations where this is used:
1) when AOTAutograd creates FakeTensor grad_outputs, it recursively calls .view_as(self) on every inner dense tensor via a transform_subclass. In this case, yes, we want to add the new nested int to the same equiv set.
2) when AOTAutograd does functionalization, it calls transform_subclass to swap all FakeTensors with FunctionalTensors, we want a different equiv set in this case because we don't want the FunctionalTensors to share the same equiv set cache as FakeTensors.

soulitzer commented on torch/_dynamo/source.py:
```diff
@@ -227,6 +227,25 @@ def name(self):
+        ]
+
+    def guard_source(self):
+        return self.base.guard_source()
```
Whoops this is on my todo

soulitzer commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -267,20 +267,26 @@ def from_real_tensor(
             # invariant is that make_meta_t only calls factories
             # for which it is not strictly necessary to use the
             # invocation manager (I think!)
+            if orig_t is not None:
+                maybe_memo = self._get_memo(orig_t)
+                if maybe_memo is not None:
+                    return maybe_memo
```
This is necessary so that if we already fakify the tensor on the nested int, we don't fakify the tensor a second time when we fakify the inner tensors during subclass fakification. Previously the fake tensor creation callback used in subclass fakification did not actually check the FakeTensor memo  (I would've expected it to, but I guess we just never needed to do that, and now that I think about this would've mean that if we had two NestedTenosr as input, their inner offsets wouldn't share FakeTensors? - I haven't checked)

ezyang commented:
So can we have a longer PR description please

ezyang commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -3,23 +3,224 @@
+    # can be associated with a vec. vec are then grouped into equivalence sets
+    # so that comparisons between the vec's nested ints can be made by comparing
+    # their equiv set. Morally two vec are in the same equiv set if and only if
+    # they have the same data, e.g., vec.cpu() and vec.cuda() belong to the same
```
This doesn't seem to match what we discussed in composability sync. @cpuhrsch's proposal, which we accepted in composability, was that morally, the size tensor is canonically a **cpu tensor**, which may have a cached cuda tensor attached to it (which we use to avoid HtoD transfer when we're running kernels.) If CUDA let us transmit as many integer parameters in CUDA launch, we would have done that instead.

In this regime, it doesn't make sense to me talk about two vecs of differing devices being in the same set. I'm never creating a nested int with a cuda tensor. I only ever have nested ints of cpu tensors (and then mumble mumble cuda caching). This has implications on public facing NT API: you can't create a NT from a CUDA tensor sizes *only* (a DtoH sync must happen if you do), but *this makes sense*, because how are you going to check if this has the same size as something else if you don't do this sync!

soulitzer replied:
> You can't create a NT from a CUDA tensor sizes only (a DtoH sync must happen if you do)

Ah I might have missed the discussion on the trade offs with this implication in the composability meeting, so after realizing that this was a new implication, I ended up convincing myself to go the other way. Here's what I understand the trade off to be:

Pros of "morally a cpu tensor"
- Avoid the implicit DtoH sync when doing `zeros([j0])
Cons:
- Adding a sync for when we create NT if offsets passed is CUDA. 

I think the cons outweigh the pros:
- This DtoH sync is very common in the use cases we're working with (some kind of torch rec JaggedTensor with lengths on CUDA) and will continue to be common until NT is used everywhere.
- The NT(offsets_on_cuda) DtoH sync isn't actually completely explicit - sure it makes sense under the  "morally a cpu tensor" framework, but users won't know that unless they study NT docs
- Implicit DtoH sync is rare - it only occurs when someone does calls a factory function where they intend for the result to be a cpu tensor.
- Implicit DtoH sync case can be mitigated by adding a warning or something.

ezyang replied:
> Adding a sync for when we create NT if offsets passed is CUDA

Well, if this is the tall pole in the tent, one could imagine lazily populating the CPU tensor in the CUDA -> CPU case, in case you never end up actually needing to consult it. I think it is at least worth thinking about this more, because there is a big pro: size equality follows intuitive rules: two tensors compare equal if their nested sizes logically compare equal. No "well, this CPU nested size is never equal to a CUDA size", no concept of device for nested size.

By the way, while JT "apparently" doesn't hold its sizes on CPU, it will happily do tons of DtoH syncs on some things on its API:

```
    def to_dense(self) -> List[torch.Tensor]:
        tensor_list = []
        for index in range(self.offsets().size(0) - 1):
            offset = self.offsets()[index].item()
            next_offset = self.offsets()[index + 1].item()
            tensor_list.append(self.values()[offset:next_offset])
        return tensor_list
```

KJT is even more interesting: we can think of it as a NT with two dimensions before the jagged dimension (feature, batch). The sizes on the batch dimension are not cached on CPU (same as NT). But the sizes on feature dimension are!

```
        values: torch.Tensor = [V0, V1, V2, V3, V4, V5, V6, V7]  # V == any tensor datatype
        weights: torch.Tensor = [W0, W1, W2, W3, W4, W5, W6, W7]  # W == any tensor datatype
        lengths: torch.Tensor = [2, 0, 1, 1, 1, 3]  # representing the jagged slice
        offsets: torch.Tensor = [0, 2, 2, 3, 4, 5, 8]  # offsets from 0 for each jagged slice
        keys: List[str] = ["Feature0", "Feature1"]  # correspond to each value of dim_0
        index_per_key: Dict[str, int] = {"Feature0": 0, "Feature1": 1}  # index for each key
        offset_per_key: List[int] = [0, 3, 8]  # start offset for each key and final offset
```

So one cannot even say torchrec isn't doing this, it is, when it is extremely helpful.

soulitzer replied:
> Well, if this is the tall pole in the tent, one could imagine lazily populating the CPU tensor in the CUDA -> CPU case, in case you never end up actually needing to consult it. I think it is at least worth thinking about this more, because there is a big pro: size equality follows intuitive rules: two tensors compare equal if their nested sizes logically compare equal. No "well, this CPU nested size is never equal to a CUDA size", no concept of device for nested size.

Ahh this is definitely my fault for procrastinating actually fleshing out the device handling code lol, but if I am understanding correctly, in the current impl, we DO obey those rules around shape equality (since vec_cpu and vec_cuda are in the same equiv set), and the intention IS actually to do the lazily population (because we want to avoid doing .to() to populate that cpu tensor immediately, and we still need some tensor placeholder to put in the weak tensor key to equiv set map, our only choice is to declare the cuda tensor as canonical). 

To be clear, all canonical means here is: all alive nested ints objects in that equiv set must point to the same vec, and that vec we call the canonical vec. The idea that a canonical vec exists at all is intended to entirely be an implementation detail. It should not matter to users whether that canonical vec is a cpu vec or a cuda vec. Concretely, if I constructed a NJT with cuda offsets, that cuda offsets would become the canonical vec purely because it is the first vec for bookkeeping reasons, but morally, the nested int returned is still cpu. That is, if I do `nt_zeros = zeros([j0])` that would create a cpu tensor. In fact, because every equivalence set also has a single canonical nested int, nt_zeros (on cpu) would hold the same exact nested int object (which still references the canonical vec which is on cuda) as the original cuda NT the nested int was extracted from.

Anyway, initially I had imagined our concern to be: if we're saying that size is device agnostic, then it might be unexpected if torch.zeros(nt_cuda.shape) incurred a implicit DtoH sync. But now if we're okay with saying: "we still have a morally a cpu tensor and the reason we incurred a DtoH sync there is that the cpu offsets is materialized lazily", that sounds a lot like a good argument to stick with the design in this PR today!

ezyang replied:
So this is where... you need an actual PR description... that says what the updated design is lol.

Based on your description here, I still have more questions:

* If two CUDA sizes tensors happen to be equal, but we have to actually perform an equality test to find out this is the case, if I test equality between these two vecs, do you discover that this is the case and merge their equivalence classes?
* What is the ownership relationship between the canonical vec (tensor) and all the other vec (tensors) in the equivalence class? Notably, if I repeatedly cast a cpu vec to cuda, do I get the same cuda vec or different cuda vecs

You don't need to implement the code to figure it out. Spend some time thinking about it and writing out an English spec.

soulitzer replied:
> So this is where... you need an actual PR description... that says what the updated design is lol.

Totally agree :P.

> If two CUDA sizes tensors happen to be equal, but we have to actually perform an equality test to find out this is the case, if I test equality between these two vecs, do you discover that this is the case and merge their equivalence classes?

I don't plan to implement merging sets in this PR, but that's certainly a possibility. In the near term, the user is expected to tell us what equiv set a vec belongs to upon registeration via `maybe_create(equiv_set_from=)`. APIs like .to() will automatically use that API. 

> What is the ownership relationship between the canonical vec (tensor) and all the other vec (tensors) in the equivalence class? Notably, if I repeatedly cast a cpu vec to cuda, do I get the same cuda vec or different cuda vecs

Same cuda vec, the equiv set dict will have dedicated "slots": vec_cuda and vec_cpu  that are intended to store cached cuda or cpu copies of the vec. If someone repeatedly casts a nt to a device, we would materialize the offset in the slot of the associated the first time, and return the same tensor the next time. Ownership-wise as long as one vec in the equiv_set is alive, the equiv set's associated meta cache will stay alive. 

This diagram might be helpful:
```
#  symint       vec         equiv_set       cached metadata
# --------------------------------------------------------------------------
#  2*j0 -----\                        /---- weak canonical vec0, weak canonical j0_0
#             v                      | /--> id: int = 0
#  j0_0 ------> vec0 -----> equiv_set0 ---> sum_vec: Optional[SymInt]
#  j0_1 ------^         /-^            \--> vec_{cuda,cpu}: Optional[Tensor]
#                      /
#               vec1--/               /---- weak vec2, weak j1_0
#                                    | /--> id: int = 1
#  j1_0 ------> vec2 -----> equiv_set1 ---> ...
```

ezyang reviewed (changes requested):
This doesn't seem to implement what was discussed in composability

soulitzer committed:
Update on "[NJT] Store vec on nested ints"


This PR:
- adds data structure managing vec and nested int as well as metadata 
- recursively fakify tensors on symint

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Store vec on nested ints"


This PR:
- adds data structure managing vec and nested int as well as metadata 
- recursively fakify tensors on symint

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Store vec on nested ints"


This PR:
- adds data structure managing vec and nested int as well as metadata 
- recursively fakify tensors on symint

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Store vec on nested ints"


This PR:
- adds data structure managing vec and nested int as well as metadata 
- recursively fakify tensors on symint

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
